Add CERTIFIED column to package list

janisz opened the pull request:
To distinguish community and regular packages, add `CERTIFIED` column to `dcos package list` command and change `SELECTED` to `CERTIFIED` in `dcos package search`.

Allow user to filter by part of the name passed as the first argument.

Fixes:
* https://jira.mesosphere.com/browse/DCOS-55773
* https://jira.mesosphere.com/browse/DCOS_OSS-5607

janisz committed:
Add selected column to package list

To distinguish community and regular packages, add `SELECTED` column to package list the same as package search`.

Fixes: https://jira.mesosphere.com/browse/DCOS-55773

makkes commented on pkg/cmd/pkg/package_list.go:
```diff
@@ -133,7 +134,7 @@ func newCmdPackageList(ctx api.Context) *cobra.Command {
 			}
 
-			table := cli.NewTable(ctx.Out(), []string{"NAME", "VERSION", "APP", "COMMAND", "DESCRIPTION"})
+			table := cli.NewTable(ctx.Out(), []string{"NAME", "VERSION", "SELECTED", "APP", "COMMAND", "DESCRIPTION"})
```
What about taking the opportunity and add a unit test for this cmd?

janisz replied:
This changes scope of this task

makkes commented on pkg/cmd/pkg/package_list.go:
```diff
@@ -133,7 +134,7 @@ func newCmdPackageList(ctx api.Context) *cobra.Command {
 			}
 
-			table := cli.NewTable(ctx.Out(), []string{"NAME", "VERSION", "APP", "COMMAND", "DESCRIPTION"})
+			table := cli.NewTable(ctx.Out(), []string{"NAME", "VERSION", "SELECTED", "APP", "COMMAND", "DESCRIPTION"})
```
After @bamarni chimed in I would prefer calling this column `CERTIFIED` which better explains its purpose.

janisz replied:
Fixed

janisz committed:
Change SELECTED to CERTIFIED

UI is using Certified to identify community pacakges.
Previously in dcos search we used SELECTED as Cosmos does.
This commit changes it to make it clear for a users that selected
has nothng about with their decision.

janisz committed:
Add missing filter

Fixes: https://jira.mesosphere.com/browse/DCOS_OSS-5607

janisz committed:
Code refactoring

janisz committed:
Add test with mocked Cosmos Client

janisz committed:
Revert test file change

janisz committed:
Add tests

makkes reviewed (approved):
awesome

drewvanstone commented on pkg/cmd/pkg/package_list_test.go:
```diff
@@ -0,0 +1,177 @@
+	client.On("PackageList").Return(packages, nil)
+
+
+	var testCases = []struct {
+		cluster string
+		options listOptions
+		out     string
+	}{
+		{NoPlugins, listOptions{jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"},{"command":{"name":"xyz"},"description":"XYZ","framework":false,"name":"package-2","selected":false,"version":"0.0.1"}]`},
+		{NoPlugins, listOptions{query: "package-2", jsonOutput: true}, `[{"command":{"name":"xyz"},"description":"XYZ","framework":false,"name":"package-2","selected":false,"version":"0.0.1"}]`},
+		{NoPlugins, listOptions{query: "-2", jsonOutput: true}, `[{"command":{"name":"xyz"},"description":"XYZ","framework":false,"name":"package-2","selected":false,"version":"0.0.1"}]`},
+		{NoPlugins, listOptions{query: "package", jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"},{"command":{"name":"xyz"},"description":"XYZ","framework":false,"name":"package-2","selected":false,"version":"0.0.1"}]`},
+		{NoPlugins, listOptions{query: "a", jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"},{"command":{"name":"xyz"},"description":"XYZ","framework":false,"name":"package-2","selected":false,"version":"0.0.1"}]`},
+		{NoPlugins, listOptions{query: "some-app", jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"}]`},
+		{NoPlugins, listOptions{appID: "some-app", jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"}]`},
+		{NoPlugins, listOptions{appID: "a", jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"}]`},
+		{NoPlugins, listOptions{appID: "b", jsonOutput: true}, `[{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"}]`},
+		{NoPlugins, listOptions{appID: "package-2", jsonOutput: true}, `[{"command":{"name":"xyz"},"description":"XYZ","framework":false,"name":"package-2","selected":false,"version":"0.0.1"}]`},
+		{NoPlugins, listOptions{cliOnly: true, jsonOutput: true}, `[]`},
+		{NoPlugins, listOptions{appID: "not found", jsonOutput: true}, `[]`},
+		{NoPlugins, listOptions{query: "not found", jsonOutput: true}, `[]`},
+		{MultipleCommands, listOptions{cliOnly: true, jsonOutput:true}, `[{"apps":["/cli-app"],"command":{"name":"cli-app"},"description":"Some CLI APP","framework":true,"name":"cli-app","selected":true,"version":"alpha"},{"apps":["/pkg-cli"],"command":{"name":"pkg-cli"},"description":"Some CLI Package","framework":true,"name":"pkg-cli","selected":true,"version":"2.4.4-1.15.4"}]`},
+		{MultipleCommands, listOptions{query: "cli", cliOnly:true, jsonOutput:true}, `[{"apps":["/cli-app"],"command":{"name":"cli-app"},"description":"Some CLI APP","framework":true,"name":"cli-app","selected":true,"version":"alpha"},{"apps":["/pkg-cli"],"command":{"name":"pkg-cli"},"description":"Some CLI Package","framework":true,"name":"pkg-cli","selected":true,"version":"2.4.4-1.15.4"}]`},
+		{MultipleCommands, listOptions{query: "app", jsonOutput:true}, `[{"apps":["/cli-app"],"command":{"name":"cli-app"},"description":"Some CLI APP","framework":true,"name":"cli-app","selected":true,"version":"alpha"},{"apps":["a","b","some-app"],"description":"package-2","framework":false,"name":"package-1","selected":false,"version":"0.0.0.1"}]`},
+		{MultipleCommands, listOptions{appID: "app", jsonOutput:true}, `[]`},
+		{MultipleCommands, listOptions{appID: "cli-app", jsonOutput:true}, `[{"apps":["/cli-app"],"command":{"name":"cli-app"},"description":"Some CLI APP","framework":true,"name":"cli-app","selected":true,"version":"alpha"}]`},
+		{MultipleCommands, listOptions{query: "package", cliOnly:true, jsonOutput:true}, `[]`},
+	}
+	for _, tt := range testCases {
+		t.Run(fmt.Sprintf("%s_%#v", tt.cluster, tt.options), func(t *testing.T) {
+			out, ctx := setupCluster(tt.cluster)
+
+			err := listPackages(ctx, tt.options, client)
+			assert.NoError(t, err)
+			assert.JSONEq(t, tt.out, out.String(), out.String())
+		})
+	}
```
![](https://media1.giphy.com/media/UlexC9HXTiNz2/giphy.gif?cid=790b7611aab07b2a7c699e847cdee15d6b5639c369cf2c13&rid=giphy.gif)

drewvanstone commented on pkg/cmd/pkg/package_list.go:
```diff
@@ -19,149 +21,175 @@ import (
+			i := strings.Index(description, "\n")
+			description = description[:i] + "..."
+		}
+		if utf8.RuneCountInString(description) > 70 {
+			description = description[0:66] + "..."
+		}
```
I like this, very clean 👍

drewvanstone commented on pkg/cmd/pkg/package_list.go:
```diff
@@ -19,149 +21,175 @@ import (
+		},
+	}
+	cmd.Flags().StringVar(&opts.appID, "app-id", "", "The application ID")
+	cmd.Flags().BoolVar(&opts.cliOnly, "cli", false, "Command line only")
```
Hmm I think the description here should be a little clearer. What does "Command line only" mean? Does it disable some web interface or something?

drewvanstone commented on pkg/cmd/pkg/package_list.go:
```diff
@@ -19,149 +21,175 @@ import (
+	}
+	cmd.Flags().StringVar(&opts.appID, "app-id", "", "The application ID")
+	cmd.Flags().BoolVar(&opts.cliOnly, "cli", false, "Command line only")
+	cmd.Flags().BoolVar(&opts.jsonOutput, "json", false, "Print options json format")
```
Wording here can be cleaned up a bit too. Maybe "Set output format to json"?

drewvanstone reviewed (changes requested):
Just a couple small points, but overall this looks fantastic!

janisz commented:
Flags descriptions should be consistent across whole project. How about fixing it in new PR?

```go
pkg/cmd/pkg/package_describe.go:        cmd.Flags().BoolVar(&opts.cliOnly, "cli", false, "Command line only")
pkg/cmd/pkg/package_install.go: cmd.Flags().BoolVar(&options.cliOnly, "cli", false, "Command line only")
pkg/cmd/pkg/package_list.go:    cmd.Flags().BoolVar(&opts.cliOnly, "cli", false, "Command line only")
pkg/cmd/pkg/package_uninstall.go:       cmd.Flags().BoolVar(&cliOnly, "cli", false, "Command line only")
```
```go
pkg/cmd/diagnostics/diagnostics_list.go:        cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/job/job_history.go:     cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/job/job_list.go:        cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/job/job_queue.go:       cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/job/job_schedule_show.go:       cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/job/job_show_runs.go:   cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_diagnostics_cancel.go:        cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_diagnostics_list.go:  cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_diagnostics_status.go:        cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_dns.go:       cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_list.go:      cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_list_components.go:   cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_metrics_details.go:   cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/node/node_metrics_summary.go:   cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/pkg/package_list.go:    cmd.Flags().BoolVar(&opts.jsonOutput, "json", false, "Print options json format")
pkg/cmd/pkg/package_repo_list.go:       cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/pkg/package_search.go:  cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/quota/quota_get.go:     cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/quota/quota_list.go:    cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/service/service_list.go:        cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/task/task_list.go:      cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/task/task_metrics_details.go:   cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
pkg/cmd/task/task_metrics_summary.go:   cmd.Flags().BoolVar(&jsonOutput, "json", false, "Print in json format")
```

makkes commented:
@mesosphere-ci run all integration tests

janisz committed:
Decrease min-complexity check

janisz committed:
Add tests for dcos package search

janisz commented:
@mesosphere-ci run linux integration tests
